zaino-state/src/chain_index/source.rs

```diff
@@ -328,8 +608,122 @@ pub(crate) mod test {
             id: Hash,
         ) -> BlockchainSourceResult<(Option<(sapling::Root, u64)>, Option<(orchard::Root, u64)>)>
         {
-            let index = self.hashes.iter().position(|h| h == &id);
-            Ok(index.map(|i| self.roots[i]).unwrap_or((None, None)))
+            let active_tip = self.active_height() as usize; // serve up to active tip
+
+            if let Some(index) = self.hashes.iter().position(|h| h == &id) {
+                if index <= active_tip {
+                    Ok(self.roots[index])
+                } else {
+                    Ok((None, None))
+                }
+            } else {
+                Ok((None, None))
+            }
+        }
+
+        async fn get_mempool_txids(
+            &self,
+        ) -> BlockchainSourceResult<Option<Vec<zebra_chain::transaction::Hash>>> {
+            let mempool_index = self.active_height() as usize + 1;
+
+            let txids = if mempool_index < self.blocks.len() {
+                self.blocks[mempool_index]
+                    .transactions
+                    .iter()
+                    .map(|transaction| transaction.hash())
+                    .collect::<Vec<_>>()
+            } else {
+                Vec::new()
+            };
+
+            Ok(Some(txids))
+        }
+
+        async fn get_transaction(
+            &self,
+            txid: Hash,
+        ) -> BlockchainSourceResult<Option<Arc<zebra_chain::transaction::Transaction>>> {
+            let txid_tr: zebra_chain::transaction::Hash =
+                zebra_chain::transaction::Hash::from(txid.0);
+
+            let active_height = self.active_height() as usize;
+            let mempool_index = active_height + 1;
+
+            for index in 0..=active_height {
+                if index >= self.blocks.len() {
+                    break;
+                }
+                if let Some(found) = self.blocks[index]
+                    .transactions
+                    .iter()
+                    .find(|transaction| transaction.hash() == txid_tr)
+                {
+                    return Ok(Some(Arc::clone(found)));
+                }
+            }
+
+            if mempool_index < self.blocks.len() {
+                if let Some(found) = self.blocks[mempool_index]
+                    .transactions
+                    .iter()
+                    .find(|transaction| transaction.hash() == txid_tr)
+                {
+                    return Ok(Some(Arc::clone(found)));
+                }
+            }
+
+            Ok(None)
+        }
+
+        async fn get_best_block_hash(
+            &self,
+        ) -> BlockchainSourceResult<Option<zebra_chain::block::Hash>> {
+            let active_tip = self.active_height() as usize;
+
+            if self.blocks.is_empty() || active_tip >= self.blocks.len() {
+                return Ok(None);
+            }
+
+            Ok(Some(self.blocks[active_tip].hash()))
```

Comment:
its best to use `get` and return errors for safety, otherwise it may panic unless its guaranteed to always find the value. also its off by one:
```suggestion
            Ok(Some(self.blocks[active_height - 1].hash()))
```

Reply:
as this is only test code i think this is ok.